### PR TITLE
chore: add procedure calls metric collector

### LINKED ISSUE #229: COUNTER: `procedure_call_total` { procedure, code }

### PR DESCRIPTION
**Pull request by Julieta11**
Fixes #229 
```j son
# HELP dcl_social_service_rpc_procedure_call Social Service RPC Websocket Procedure Calls
# TYPE dcl_social_service_rpc_procedure_call counter
dcl_social_service_rpc_procedure_call{procedure="GetFriends",code="OK"} 1
dcl_social_service_rpc_procedure_call{procedure="GetRequestEvents",code="OK"} 1
dcl_social_service_rpc_procedure_call{procedure="SubscribeFriendshipEventsUpdates",code="OK"} 1
dcl_social_service_rpc_procedure_call{procedure="UpdateFriendshipEvent",code="INTERNAL_SERVER_ERROR"} 2
```

### CODE REVIEW

File: src/ws/metrics.rs [hunk 1]
```diff
@@ -28,28 +28,28 @@ impl Procedure {
...
-        WsServiceError::BadRequest(_) => "BAD_REQUEST",
-        WsServiceError::Forbidden(_) => "FORBIDDEN",
-        WsServiceError::TooManyRequests(_) => "TOO_MANY_REQUESTS",
+pub fn record_procedure_call(code: Option<WsServiceError>, procedure: Procedure) {
+    let label = match code {
+        Some(WsServiceError::Unauthorized(_)) => "UNAUTHORIZED",
+        Some(WsServiceError::InternalServer(_)) => "INTERNAL_SERVER",
+        Some(WsServiceError::BadRequest(_)) => "BAD_REQUEST",
+        Some(WsServiceError::Forbidden(_)) => "FORBIDDEN",
+        Some(WsServiceError::TooManyRequests(_)) => "TOO_MANY_REQUESTS",
```
Comment (agusaldasoro) [inline src/ws/metrics.rs:37]:
```suggestion
        Some(WsServiceError::Unauthorized(_)) => "UNAUTHORIZED_ERROR",
        Some(WsServiceError::InternalServer(_)) => "INTERNAL_SERVER_ERROR",
        Some(WsServiceError::BadRequest(_)) => "BAD_REQUEST_ERROR",
        Some(WsServiceError::Forbidden(_)) => "FORBIDDEN_ERROR",
        Some(WsServiceError::TooManyRequests(_)) => "TOO_MANY_REQUESTS_ERROR",
```

File: src/ws/metrics.rs [hunk 2]
```diff
@@ -28,28 +28,28 @@ impl Procedure {
...
 
-pub fn record_error_response_code(error: WsServiceError, procedure: Procedure) {
-    let label = match error {
-        WsServiceError::Unauthorized(_) => "UNAUTHORIZED",
-        WsServiceError::InternalServer(_) => "INTERNAL_SERVER",
-        WsServiceError::BadRequest(_) => "BAD_REQUEST",
-        WsServiceError::Forbidden(_) => "FORBIDDEN",
-        WsServiceError::TooManyRequests(_) => "TOO_MANY_REQUESTS",
+pub fn record_procedure_call(code: Option<WsServiceError>, procedure: Procedure) {
+    let label = match code {
```
Comment (agusaldasoro) [inline src/ws/metrics.rs:32]:
```suggestion
    let code = match code {
```

File: src/ws/metrics.rs [hunk 3]
```diff
@@ -99,14 +99,14 @@ pub async fn validate_bearer_token(
 }
 
 lazy_static! {
-    pub static ref ERROR_RESPONSE_CODE_COLLECTOR: IntCounterVec = {
+    pub static ref PROCEDURE_CALL_COLLECTOR: IntCounterVec = {
         let opts = Opts::new(
-            "dcl_social_service_rpc_error_response_code",
-            "Social Service RPC Websocket Error Response Codes",
+            "dcl_social_service_rpc_procedure_call",
```
Comment (agusaldasoro) [inline src/ws/metrics.rs:104]:
```suggestion
            "ws_rpc_procedure_call",
```

**Review by agusaldasoro — CHANGES_REQUESTED** [review]
Great work! Can we change the metrics to:
- Be more generic so the name of the metric doesn't include the service, but it's a tag
- Include the word "total" as it is a counter
- Add the `_ERROR` ending for errors, so we can filter them out with a regex

Something like:

```
# HELP ws_rpc_procedure_call_total Social Service RPC Websocket Procedure Calls
# TYPE ws_rpc_procedure_call_total counter
ws_rpc_procedure_call_total{service=social-service,procedure="GetFriends",code="OK"} 1
ws_rpc_procedure_call_total{service=social-service,procedure="GetRequestEvents",code="OK"} 1
ws_rpc_procedure_call_total{service=social-service,procedure="SubscribeFriendshipEventsUpdates",code="OK"} 1
ws_rpc_procedure_call_total{service=social-service,procedure="UpdateFriendshipEvent",code="INTERNAL_SERVER"} 2
```

File: src/ws/metrics.rs (hunk 3 above)
Comment (agusaldasoro) [reply src/ws/metrics.rs:104]: mm, okay, but aren't those for business logic metrics only?

Comment (Julieta11) [reply src/ws/metrics.rs:104]: I don't know, there is no such classification in the docs. We use the same naming convention for the current REST API: 
+ `dcl_social_service_http_requests_duration_seconds_bucket`
+ `dcl_social_service_http_requests_duration_seconds_count`
+ `dcl_social_service_http_requests_duration_seconds_sum`
+ `dcl_social_service_http_requests_total`

**Review by lauti7 — COMMENTED** [review]
Just one comment. It would be great If we can avoid using lazy_static.

File: src/ws/metrics.rs [hunk 4]
```diff
@@ -99,14 +99,14 @@ pub async fn validate_bearer_token(
...
+            "Social Service RPC Websocket Procedure Calls",
         );
 
-        IntCounterVec::new(opts, &["status_code", "procedure"])
-            .expect("dcl_social_service_rpc_error_response_code metric can be created")
+        IntCounterVec::new(opts, &["code", "procedure"])
+            .expect("dcl_social_service_rpc_procedure_call metric can be created")
```
[the remark this answers is not in the document]
Comment (Julieta11) [reply src/ws/metrics.rs]: Moved to the context 💯

File: src/ws/metrics.rs [hunk 5]
```diff
@@ -1,10 +1,43 @@
...
+impl Default for Metrics {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Metrics {
+    pub fn new() -> Self {
+        let opts = Opts::new(
+            "dcl_social_service_rpc_procedure_call",
```
Comment (agusaldasoro) [inline src/ws/metrics.rs:25]:
```suggestion
            "dcl_social_service_rpc_procedure_call_total",
```

File: src/ws/metrics.rs [hunk 6]
```diff
@@ -1,10 +1,43 @@
...
+            "Social Service RPC Websocket Procedure Calls",
+        );
+
+        let procedure_call_collector = IntCounterVec::new(opts, &["code", "procedure"])
+            .expect("dcl_social_service_rpc_procedure_call metric can be created");
```
Comment (agusaldasoro) [inline src/ws/metrics.rs:30]: Shouldn't this be the message when something fails?

Comment (Julieta11) [reply src/ws/metrics.rs:30]: There are[ two styles](https://doc.rust-lang.org/stable/std/error/index.html#common-message-styles) for the expect messages and the recommended style is the “expect as precondition” style. I thought I was using the recommended style but clearly I'm not 🤣. I'll update it to a more a suitable one.

**Review by agusaldasoro — APPROVED** [review]

**Pull request merged**